# Housing Rental Analysis for San Francisco

In [16]:
# Import the required libraries and dependencies
import pandas as pd
import hvplot.pandas
from pathlib import Path
import geoviews as gv

## Import the data 

In [17]:
# Importing the sfo_neighborhoods_census_data.csv
sfo_data_df = pd.read_csv(
    Path('Resources/sfo_neighborhoods_census_data.csv')   
).drop_duplicates()

# Review the first and last five rows of the DataFrame
display(sfo_data_df.head())
display(sfo_data_df.tail())

year      neighborhood  sale_price_sqr_foot  housing_units  gross_rent
0  2010      Alamo Square           291.182945         372560        1239
1  2010        Anza Vista           267.932583         372560        1239
2  2010           Bayview           170.098665         372560        1239
3  2010  Buena Vista Park           347.394919         372560        1239
4  2010  Central Richmond           319.027623         372560        1239

year            neighborhood  sale_price_sqr_foot  housing_units  \
392  2016          Telegraph Hill           903.049771         384242   
393  2016              Twin Peaks           970.085470         384242   
394  2016  Van Ness/ Civic Center           552.602567         384242   
395  2016       Visitacion Valley           328.319007         384242   
396  2016           Westwood Park           631.195426         384242   

     gross_rent  
392        4390  
393        4390  
394        4390  
395        4390  
396        4390

---

## Calculate and Plot the Housing Units per Year

In [18]:
# Create a numerical aggregation that groups the data by the year and then averages the results.
housing_units_by_year = (
    sfo_data_df
    .groupby('year')
    .mean()
)

display(housing_units_by_year)

sale_price_sqr_foot  housing_units  gross_rent
year                                                
2010           369.344353       372560.0      1239.0
2011           341.903429       374507.0      1530.0
2012           399.389968       376454.0      2324.0
2013           483.600304       378401.0      2971.0
2014           556.277273       380348.0      3528.0
2015           632.540352       382295.0      3739.0
2016           697.643709       384242.0      4390.0

In [19]:
# Create a visual aggregation explore the housing units by year
housing_units_by_year.hvplot.bar(
    x='year',
    y='housing_units',
    ylim=(365000, 385000),
    title='Housing Units in San Francisco 2010 - 2016',
    ).opts(
    yformatter='%.0f')

:Bars   [year]   (housing_units)

The overall trend in housing_units over the period being analyzed is growth. Every year from 2010 to 2016 there are more housing units in San Francisco. There is not a year when there was a loss of units or even a maintenence of units, there is consistent growth year after year.

---

## Calculate and Plot the Average Sale Prices per Square Foot

In [20]:
# Create a numerical aggregation by grouping the data by year and averaging the results
prices_square_foot_by_year = (
    sfo_data_df
    .groupby('year')
    .mean()
)

display(prices_square_foot_by_year)


sale_price_sqr_foot  housing_units  gross_rent
year                                                
2010           369.344353       372560.0      1239.0
2011           341.903429       374507.0      1530.0
2012           399.389968       376454.0      2324.0
2013           483.600304       378401.0      2971.0
2014           556.277273       380348.0      3528.0
2015           632.540352       382295.0      3739.0
2016           697.643709       384242.0      4390.0

The lowest gross rent reported for the years included in the DataFrame is $1239 in  2010

In [21]:
# Filter out the housing_units column, creating a new DataFrame 
# Keep only sale_price_sqr_foot and gross_rent averages per year
prices_square_foot_by_year = (
    sfo_data_df.drop(columns=['housing_units'])
    .groupby('year')
    .mean()
)

display(prices_square_foot_by_year)

sale_price_sqr_foot  gross_rent
year                                 
2010           369.344353      1239.0
2011           341.903429      1530.0
2012           399.389968      2324.0
2013           483.600304      2971.0
2014           556.277273      3528.0
2015           632.540352      3739.0
2016           697.643709      4390.0

In [23]:
# Plot prices_square_foot_by_year. 
prices_square_foot_by_year.hvplot(
    title = 'Sale Price per Sq Ft and Avg Gross Rent - San Francisco 2010 - 2016',
    xlabel = 'Year',
    ylabel = 'Gross Rent/Sale Price per Sq Foot'
)

:NdOverlay   [Variable]
   :Curve   [year]   (value)

The average sale price per square foot in 2011 was 27.440924 LESS than the average sale price per square foot in 2010.

Despite the drop in sale price per square foot in 2011, the gross rent in 2011 actually increased over the 2010 average gross rent.

---

## Compare the Average Sale Prices by Neighborhood

In [24]:
# Group by year and neighborhood and then create a new dataframe of the mean values
prices_by_year_by_neighborhood = (
    sfo_data_df
    .groupby(['year','neighborhood'])
    .mean()
)

display(prices_by_year_by_neighborhood)

sale_price_sqr_foot  housing_units  gross_rent
year neighborhood                                                          
2010 Alamo Square                     291.182945       372560.0      1239.0
     Anza Vista                       267.932583       372560.0      1239.0
     Bayview                          170.098665       372560.0      1239.0
     Buena Vista Park                 347.394919       372560.0      1239.0
     Central Richmond                 319.027623       372560.0      1239.0
...                                          ...            ...         ...
2016 Telegraph Hill                   903.049771       384242.0      4390.0
     Twin Peaks                       970.085470       384242.0      4390.0
     Van Ness/ Civic Center           552.602567       384242.0      4390.0
     Visitacion Valley                328.319007       384242.0      4390.0
     Westwood Park                    631.195426       384242.0      4390.0

[397 rows x 3 columns]

In [25]:
# Filter out the housing_units
prices_by_year_by_neighborhood = prices_by_year_by_neighborhood.drop(columns=['housing_units'])

display(prices_by_year_by_neighborhood.head())
display(prices_by_year_by_neighborhood.tail())

sale_price_sqr_foot  gross_rent
year neighborhood                                     
2010 Alamo Square               291.182945      1239.0
     Anza Vista                 267.932583      1239.0
     Bayview                    170.098665      1239.0
     Buena Vista Park           347.394919      1239.0
     Central Richmond           319.027623      1239.0

sale_price_sqr_foot  gross_rent
year neighborhood                                           
2016 Telegraph Hill                   903.049771      4390.0
     Twin Peaks                       970.085470      4390.0
     Van Ness/ Civic Center           552.602567      4390.0
     Visitacion Valley                328.319007      4390.0
     Westwood Park                    631.195426      4390.0

In [26]:
# Create an interactive line plot of the average price per square foot

prices_by_year_by_neighborhood.hvplot(
    title = 'Sale Price per Sq Ft & Avg Gross Rent - SFO 2010 - 2016 - by Neighborhood',
    x = 'year', 
    xlabel = 'Year',
    ylabel = 'Gross Rent/Sale Price per Sq Foot',
    groupby = 'neighborhood'
)


:DynamicMap   [neighborhood]
   :NdOverlay   [Variable]
      :Curve   [year]   (value)

For the Anza Vista neighborhood, the average sale price per square foot for 2016 was 88.402, 256.089 less than the price for 2012 which was 344.491

---

## Build an Interactive Neighborhood Map

In [27]:
# Load neighborhoods coordinates data
neighborhood_locations_df = pd.read_csv(
    Path('Resources/neighborhoods_coordinates.csv'),
    index_col = 'Neighborhood'
)
# Review the DataFrame
display(neighborhood_locations_df)

Lat         Lon
Neighborhood                             
Alamo Square        37.791012 -122.402100
Anza Vista          37.779598 -122.443451
Bayview             37.734670 -122.401060
Bayview Heights     37.728740 -122.410980
Bernal Heights      37.728630 -122.443050
...                       ...         ...
West Portal         37.740260 -122.463880
Western Addition    37.792980 -122.435790
Westwood Highlands  37.734700 -122.456854
Westwood Park       37.734150 -122.457000
Yerba Buena         37.792980 -122.396360

[73 rows x 2 columns]

In [28]:
# Calculate the mean values for each neighborhood
all_neighborhood_info_df = (
    sfo_data_df
    .groupby('neighborhood')
    .mean()
)
# Review the resulting DataFrame
all_neighborhood_info_df

year  sale_price_sqr_foot  housing_units  \
neighborhood                                                          
Alamo Square        2013.000000           366.020712      378401.00   
Anza Vista          2013.333333           373.382198      379050.00   
Bayview             2012.000000           204.588623      376454.00   
Bayview Heights     2015.000000           590.792839      382295.00   
Bernal Heights      2013.500000           576.746488      379374.50   
...                         ...                  ...            ...   
West Portal         2012.250000           498.488485      376940.75   
Western Addition    2012.500000           307.562201      377427.50   
Westwood Highlands  2012.000000           533.703935      376454.00   
Westwood Park       2015.000000           687.087575      382295.00   
Yerba Buena         2012.500000           576.709848      377427.50   

                     gross_rent  
neighborhood                     
Alamo Square        2817.285714  
Anza Vista          3031.833333  
Bayview             2318.400000  
Bayview Heights     3739.000000  
Bernal Heights      3080.333333  
...                         ...  
West Portal         2515.500000  
Western Addition    2555.166667  
Westwood Highlands  2250.500000  
Westwood Park       3959.000000  
Yerba Buena         2555.166667  

[73 rows x 4 columns]

In [30]:
# Join the neighborhood_locations_df and the all_neighborhood_info_df DataFrame
# The axis of the concatenation is "columns".

all_neighborhoods_df = pd.concat(
    [neighborhood_locations_df, all_neighborhood_info_df], 
    axis="columns",
    sort=False
)

# Review the resulting DataFrame
display(all_neighborhoods_df.head())
display(all_neighborhoods_df.tail())


Lat         Lon         year  sale_price_sqr_foot  \
Alamo Square     37.791012 -122.402100  2013.000000           366.020712   
Anza Vista       37.779598 -122.443451  2013.333333           373.382198   
Bayview          37.734670 -122.401060  2012.000000           204.588623   
Bayview Heights  37.728740 -122.410980  2015.000000           590.792839   
Bernal Heights   37.728630 -122.443050          NaN                  NaN   

                 housing_units   gross_rent  
Alamo Square          378401.0  2817.285714  
Anza Vista            379050.0  3031.833333  
Bayview               376454.0  2318.400000  
Bayview Heights       382295.0  3739.000000  
Bernal Heights             NaN          NaN

Lat        Lon    year  sale_price_sqr_foot  \
Yerba Buena      37.79298 -122.39636  2012.5           576.709848   
Bernal Heights        NaN        NaN  2013.5           576.746488   
Downtown              NaN        NaN  2013.0           391.434378   
Ingleside             NaN        NaN  2012.5           367.895144   
Outer Richmond        NaN        NaN  2013.0           473.900773   

                 housing_units   gross_rent  
Yerba Buena           377427.5  2555.166667  
Bernal Heights        379374.5  3080.333333  
Downtown              378401.0  2817.285714  
Ingleside             377427.5  2509.000000  
Outer Richmond        378401.0  2817.285714

In [31]:
# Call the dropna function to remove any neighborhoods that do not have data
all_neighborhoods_df = all_neighborhoods_df.reset_index().dropna()

# Rename the "index" column as "Neighborhood" for use in the Visualization
all_neighborhoods_df = all_neighborhoods_df.rename(columns={"index": "Neighborhood"})

# Review the resulting DataFrame
display(all_neighborhoods_df.head())
display(all_neighborhoods_df.tail())

Neighborhood        Lat         Lon         year  sale_price_sqr_foot  \
0      Alamo Square  37.791012 -122.402100  2013.000000           366.020712   
1        Anza Vista  37.779598 -122.443451  2013.333333           373.382198   
2           Bayview  37.734670 -122.401060  2012.000000           204.588623   
3   Bayview Heights  37.728740 -122.410980  2015.000000           590.792839   
5  Buena Vista Park  37.768160 -122.439330  2012.833333           452.680591   

   housing_units   gross_rent  
0       378401.0  2817.285714  
1       379050.0  3031.833333  
2       376454.0  2318.400000  
3       382295.0  3739.000000  
5       378076.5  2698.833333

Neighborhood       Lat         Lon     year  sale_price_sqr_foot  \
68         West Portal  37.74026 -122.463880  2012.25           498.488485   
69    Western Addition  37.79298 -122.435790  2012.50           307.562201   
70  Westwood Highlands  37.73470 -122.456854  2012.00           533.703935   
71       Westwood Park  37.73415 -122.457000  2015.00           687.087575   
72         Yerba Buena  37.79298 -122.396360  2012.50           576.709848   

    housing_units   gross_rent  
68      376940.75  2515.500000  
69      377427.50  2555.166667  
70      376454.00  2250.500000  
71      382295.00  3959.000000  
72      377427.50  2555.166667

### Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame.

In [32]:
# Create a plot to analyze neighborhood info
all_neighborhoods_df.hvplot.points(
    'Lon',
    'Lat',
    geo = True,
    size = 'sale_price_sqr_foot',
    color = 'gross_rent',
    frame_width = 700,
    frame_height = 500,
    title = 'San Francisco Neighborhoods',
    tiles = True,
    hover_cols = 'Neighborhood'
)

:Overlay
   .Tiles.I  :Tiles   [x,y]
   .Points.I :Points   [Lon,Lat]   (gross_rent,sale_price_sqr_foot,Neighborhood)

The neighborhood has the highest gross rent appears to be Westwood Park. The neighborhood with the highest sale price per square foot appears to be Merced Heights.

## Compose Your Data Story

Rental real estate income in San Francisco over the time period analyzed, 2010 - 2016 had consistent growth across all neighborhoods. The sales price per square foot varied by neighborhood with some neighborhoods experiencing a drop from 2013 to 2016 opening up an opportunity for purchase of rental property and realize immediate gains in being able to rent the properties at higher rental amounts.

The data points to several neighborhoods in San Francisco in 2016 that had a drop in price per square foot sales but maintained a consistent growth in rental income. I would recommend watching for buy-and-rent opportunities in Alamo Square, Anza Vista, Hayes Valley, Ingelside, Inner Sunset, Lower Pacific Heights, Marina, Oceanview, Park North, Portero Hill and Outer Mission. These neighborhoods showed a drop in sales prices and an increase in rental prices.